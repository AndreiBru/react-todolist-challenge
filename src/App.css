.container {
    max-width: 800px;
    margin: 0 auto;
    font-family: "Trebuchet MS", Arial, Helvetica, sans-serif;
}

.completed {
  text-decoration: line-through;
}

.error-message {
  text-align: center;
  font-weight: bold;
  margin-top: 20px;
  color: tomato;
}
