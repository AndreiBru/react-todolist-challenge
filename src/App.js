import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchTodos, postTodo } from './actions/listActions';

import './App.css';

import List from './components/list';
import Input from './components/input';
import Button from './components/button';

export class App extends Component {
  componentDidMount() {
    const { fetchTodos } = this.props;

    fetchTodos();
  }

  state = {
    currentTodo: '',
  };

  addTodo = () => {
    const { postTodo } = this.props;

    postTodo(this.state.currentTodo);

    this.setState({
      currentTodo: '',
    });
  };

  updateCurrentTodo = todo => {
    this.setState({
      currentTodo: todo,
    });
  };

  render() {
    const { todos, errorMessage } = this.props;

    return (
      <div className="container">
        <h1>Todo List</h1>

        <div className="add-item-to-list">
          <Input
            name="item"
            placeholder="New Item..."
            value={this.state.currentTodo}
            onChange={this.updateCurrentTodo}
          />
          <Button onClick={this.addTodo} type="add">
            Add
          </Button>
        </div>

        <List todos={todos} />

        <div className="error-message">{errorMessage}</div>
      </div>
    );
  }
}

const mapStateToProps = state => state.list;

const mapDispatchToProps = { fetchTodos, postTodo };

export default connect(mapStateToProps, mapDispatchToProps)(App);
