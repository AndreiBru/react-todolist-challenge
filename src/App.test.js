import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import { App } from './App';

jest.mock('react-redux', a => ({
  connect: () => WrappedComponent => WrappedComponent
}));

const todos = [
  {
    title: 'one',
    url:
      'http://todo-backend-webpy.herokuapp.com/9291c49388e7436387954a26d9caae79',
    completed: false
  },
  {
    order: 3,
    title: 'two',
    url:
      'http://todo-backend-webpy.herokuapp.com/655d30051db2491bb61457a50a643d7f',
    completed: true
  },
  {
    title: 'three',
    url:
      'http://todo-backend-webpy.herokuapp.com/dae40261aee54292bad5f3901a9a8a61',
    completed: true
  },
  {
    title: 'four',
    url:
      'http://todo-backend-webpy.herokuapp.com/3f5228a4396f43af9f04272a94a7ce33',
    completed: false
  },
  {
    order: 2,
    title: 'five',
    url:
      'http://todo-backend-webpy.herokuapp.com/ca541adf6a3d4c54adf4779f9fa6af9e',
    completed: false
  }
];

describe('App', () => {
  it('renders and matches snapshot', () => {
    const comp = renderer.create(
      <App todos={todos} fetchTodos={Function.prototype} />
    );

    expect(comp.toJSON()).toMatchSnapshot();
  });

  it('calls the fetchTodos handler upon mount', () => {
    const fetchTodosSpy = jest.fn();
    shallow(<App todos={todos} fetchTodos={fetchTodosSpy} />);

    expect(fetchTodosSpy).toHaveBeenCalledTimes(1);
  });

  it("updates the input's text when typing", () => {
    const comp = mount(<App todos={todos} fetchTodos={Function.prototype} />);

    const newValue = 'Make dinner';
    comp.find('input').simulate('change', { target: { value: newValue } });

    expect(comp.find('input').prop('value')).toBe(newValue);
  });

  it('calls the postTodo handler when clicking the add button', () => {
    const postTodoSpy = jest.fn();
    const comp = mount(
      <App
        todos={todos}
        fetchTodos={Function.prototype}
        postTodo={postTodoSpy}
      />
    );

    const newValue = 'Make dinner';
    comp.find('input').simulate('change', { target: { value: newValue } });

    expect(comp.find('input').prop('value')).toBe(newValue);

    expect(postTodoSpy).not.toHaveBeenCalled();

    comp.find('[type="add"]').simulate('click');

    expect(postTodoSpy).toHaveBeenCalledTimes(1);
    expect(postTodoSpy).toHaveBeenCalledWith(newValue);
  });
});
