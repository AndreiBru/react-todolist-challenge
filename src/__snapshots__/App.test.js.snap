// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders and matches snapshot 1`] = `
<div
  className="container"
>
  <h1>
    Todo List
  </h1>
  <div
    className="add-item-to-list"
  >
    <input
      className={undefined}
      name="item"
      onChange={[Function]}
      placeholder="New Item..."
      type="text"
      value=""
    />
    <button
      className="button add"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <table>
    <thead>
      <tr>
        <th
          width="66%"
        >
          Item
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className={null}
        >
          one
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="completed"
        >
          two
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          className="completed"
        >
          three
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          className={null}
        >
          four
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          className={null}
        >
          five
        </td>
        <td>
          <button
            className="button edit"
            onClick={[Function]}
          >
            Edit
          </button>
          <button
            className="button delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="error-message"
  />
</div>
`;
